OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+/** \file
+* This file defines the  DataTable class, which is used by OpenSim to provide  *
+* an in-memory container for data access and manipulation.                     *
+*/
+
+#ifndef OPENSIM_DATA_TABLE_H_
+#define OPENSIM_DATA_TABLE_H_
+
+// Standard headers.
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <string>
+#include <type_traits>
+#include <limits>
+// Non-standard headers.
+#include <SimTKcommon.h>
+
+
+namespace OpenSim {
+  class AbstractDataTable;
+  template<typename ET> class DataTable_;
+
+  using DataTable = DataTable_<SimTK::Real>;
+
+  /// Enum to specify if the the input iterator is traversing data row-wise or
+  /// col-wise.
+  /// clang3.6 crashes if this is turned to a "enum class". Until it is fixed,
+  /// use a pre c++11 enum.
+  enum InputItDir{
+    ROWWISE, 
+    COLWISE
+  };
+
+  /// Bind two DataTables by row.
+  template<typename ET>
+  DataTable_<ET> rbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  /// Bind two DataTables by col.
+  template<typename ET>
+  DataTable_<ET> cbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  // Exceptions.
+  class ColumnDoesNotExist;
+  class ColumnHasLabel;
+  class ColumnHasNoLabel;
+  class ZeroElements;
+  class InvalidEntry;
+  class MetaDataKeyExists;
+  class MetaDataKeyDoesNotExist;
+  class MetaDataTypeMismatch;
+} // namespace OpenSim
+
+
+/**----------------------------------------------------------------------------*
+* AbstractDataTable is the base-class of all DataTable_ allowing storage of    *
+* DataTable_ of different types to be stored in containers. See Datatable_ for *
+* details on the interface.                                                    *
+*-----------------------------------------------------------------------------*/
+class OpenSim::AbstractDataTable {
+public:
+  AbstractDataTable() = default;
+  AbstractDataTable(const AbstractDataTable&) = delete;
+  AbstractDataTable(AbstractDataTable&&) = delete;
```

Comment:
Is this OK to delete the copy and move constructors? At least the copy constructor is needed by derived classes to implement `clone()`. I think if you don't have a base class copy constructor that will prevent generation of default copy constructors for the derived classes, since those should invoke their parent class's copy constructor. Am I wrong about that?